pkg/armrpc/asyncoperation/worker/service.go

```diff
@@ -18,64 +18,56 @@ package worker
 
 import (
 	"context"
+	"sync"
 
 	manager "github.com/radius-project/radius/pkg/armrpc/asyncoperation/statusmanager"
-	"github.com/radius-project/radius/pkg/armrpc/hostoptions"
-	"github.com/radius-project/radius/pkg/components/database/databaseprovider"
+	"github.com/radius-project/radius/pkg/components/database"
 	"github.com/radius-project/radius/pkg/components/queue"
-	"github.com/radius-project/radius/pkg/components/queue/queueprovider"
 	"github.com/radius-project/radius/pkg/ucp/ucplog"
 )
 
 // Service is the base worker service implementation to initialize and start worker.
+// All exported fields should be initialized by the caller.
 type Service struct {
-	// ProviderName is the name of provider namespace.
-	ProviderName string
-	// Options is the server hosting options.
-	Options hostoptions.HostOptions
-	// DatabaseProvider is the provider of the database client.
-	DatabaseProvider *databaseprovider.DatabaseProvider
+	// DatabaseClient is database client.
+	DatabaseClient database.Client
+
 	// OperationStatusManager is the manager of the operation status.
 	OperationStatusManager manager.StatusManager
-	// Controllers is the registry of the async operation controllers.
-	Controllers *ControllerRegistry
-	// RequestQueue is the queue client for async operation request message.
-	RequestQueue queue.Client
-}
 
-// Init initializes worker service - it initializes the StorageProvider, RequestQueue, OperationStatusManager, Controllers, KubeClient and
-// returns an error if any of these operations fail.
-func (s *Service) Init(ctx context.Context) error {
-	s.DatabaseProvider = databaseprovider.FromOptions(s.Options.Config.DatabaseProvider)
-	qp := queueprovider.New(s.Options.Config.QueueProvider)
+	// Options configures options for the async worker.
+	Options Options
 
-	var err error
-	storageClient, err := s.DatabaseProvider.GetClient(ctx)
-	if err != nil {
-		return err
-	}
+	// QueueProvider is the queue client.
+	QueueClient queue.Client
 
-	s.RequestQueue, err = qp.GetClient(ctx)
-	if err != nil {
-		return err
-	}
+	// controllers is the registry of the async operation controllers.
+	controllers *ControllerRegistry
 
-	s.OperationStatusManager = manager.New(storageClient, s.RequestQueue, s.Options.Config.Env.RoleLocation)
-	s.Controllers = NewControllerRegistry()
-	return nil
+	// controllersInit is used to ensure single initialization of controllers.
+	controllersInit sync.Once
+}
+
+// Controllers returns the controller registry for the worker service.
+func (s *Service) Controllers() *ControllerRegistry {
```

Comment:
Are we sure that this is only going to be called once? Should we check if s.controllers are not empty before calling **NewControllerRegistry()**?

Reply:
Yes. This is what `sync.Once` is for.